builtins/web/fetch/headers.cpp

```diff
@@ -622,23 +720,48 @@ bool Headers::set(JSContext *cx, unsigned argc, JS::Value *vp) {
     }
   } else {
     MOZ_ASSERT(mode == Mode::ContentOnly);
-    RootedObject entries(cx, get_entries(cx, self));
-    if (!entries) {
-      return false;
-    }
-
-    RootedValue name_val(cx);
-    if (!redecode_str_if_changed(cx, args[0], name_chars, name_changed, &name_val)) {
-      return false;
-    }
-
-    RootedValue value_val(cx);
-    if (!redecode_str_if_changed(cx, args[1], value_chars, value_changed, &value_val)) {
-      return false;
-    }
 
-    if (!MapSet(cx, entries, name_val, value_val)) {
-      return false;
+    auto idx = Headers::lookup(cx, self, name_chars);
+    if (!idx) {
+      if (!append_valid_normalized_header(cx, self, std::move(name_chars), std::move(value_chars)))
+        return false;
+    } else {
+      size_t index = idx.value();
+      // The lookup above will guarantee that sort_list is up to date.
+      std::vector<size_t> *headers_sort_list = static_cast<std::vector<size_t> *>(
+          JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersSortList))
+              .toPrivate());
+      MOZ_ASSERT(headers_sort_list);
+
+      HeadersList *headers_list = static_cast<Headers::HeadersList *>(
+          JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersList)).toPrivate());
+      MOZ_ASSERT(headers_list);
+
+      // Update the first entry in place to the new value
+      host_api::HostString *header_val =
+          &std::get<1>(headers_list->at(headers_sort_list->at(index)));
+
+      // Swap in the new value respecting the disposal semantics
+      header_val->ptr.swap(value_chars.ptr);
+      header_val->len = value_chars.len;
+
+      // Delete all subsequent entries for this header excluding the first,
+      // as a variation of Headers::delete.
+      size_t len = headers_list->size();
+      size_t delete_cnt = 0;
+      while (index + delete_cnt + 1 < len &&
+             headers_sort_list->at(index + delete_cnt + 1) >= delete_cnt &&
+             (header_compare(std::get<0>(headers_list->at(
+                                 headers_sort_list->at(index + delete_cnt + 1) - delete_cnt)),
+                             name_chars) == Ordering::Equal)) {
+        headers_list->erase(headers_list->begin() + headers_sort_list->at(index + delete_cnt + 1) -
```

Comment:
This isn't possible because we don't assume that same-value headers are contiguous.

Reply:
We could assume same-value headers are contiguous if we always perform joining when populating the headers list from the host. I only discovered late in the implementation that this is truly unobservable since passing headers to a response always runs through a re-iteration and sorting.

The observability of different orderings is then only observable in the case of `new Response().headers.append('unordered')` allowing unordered construction on the wire.

Note that this is also a justification for treating `initv` as always running through the guest iteration, since joining and sorting is an expectation here, whereas a host clone in theory would not join and sort.